File: frontend/src/types/order.ts
import type { PaginationParams } from './api';

export type OrderStatus = 'pending' | 'processing' | 'shipped' | 'delivered' | 'cancelled';

export interface ShipperInfo {
    shipper_id: number;
    company_name: string;
    phone?: string;
}

export interface CustomerInfo {
    customer_id: string;
    company_name: string;
    contact_name?: string;
}

export interface EmployeeInfo {
    employee_id: number;
    first_name: string;
    last_name: string;
}

export interface OrderDetail {
    product_id: number;
    product_name: string;
    unit_price: number;
    quantity: number;
    discount: number;
    line_total: number;
    discount_amount: number;
    final_total: number;
}

export interface Order {
    order_id: number;
    customer_id?: string; // Backend sends object but list view might use flat
    employee_id?: number;

    // Relationships
    customer: CustomerInfo;
    employee?: EmployeeInfo;
    shipper?: ShipperInfo;

    order_date: string | null;
    required_date: string | null;
    shipped_date: string | null;
    ship_via: number | null;
    freight: number;
    ship_name: string | null;
    ship_address: string | null;
    ship_city: string | null;
    ship_region: string | null;
    ship_postal_code: string | null;
    ship_country: string | null;
    status: OrderStatus;

    subtotal: number;
    discount_total: number;
    total: number;

    order_details: OrderDetail[];
}

export interface OrderListResponse {
    order_id: number;
    customer?: CustomerInfo;
    employee?: EmployeeInfo;
    order_date: string | null;
    required_date: string | null;
    shipped_date: string | null;
    status: OrderStatus;
    total: number;
}

export interface StartOrderDetail {
    product_id: number;
    quantity: number;
    unit_price?: number; // Optional override
    discount: number;
}

export interface CreateOrderData {
    customer_id: string;
    employee_id?: number | null;
    order_date?: string | null;
    required_date?: string | null;
    ship_via?: number | null;
    freight?: number;
    ship_name?: string;
    ship_address?: string;
    ship_city?: string;
    ship_region?: string;
    ship_postal_code?: string;
    ship_country?: string;
    order_details: StartOrderDetail[];
}

export interface UpdateOrderData {
    order_date?: string | null;
    required_date?: string | null;
    shipped_date?: string | null;
    ship_via?: number | null;
    freight?: number;
    ship_name?: string;
    ship_address?: string;
    ship_city?: string;
    ship_region?: string;
    ship_postal_code?: string;
    ship_country?: string;
}

export interface OrderFilterParams extends PaginationParams {
    status?: OrderStatus;
    customer_id?: string;
    employee_id?: number;
    date_from?: string;
    date_to?: string;
    sort_by?: string;
    sort_order?: 'asc' | 'desc';
}

File: backend/app/schemas/order.py
from pydantic import BaseModel, Field, ConfigDict
from typing import Optional, List
from datetime import date
from decimal import Decimal
from app.models.order import OrderStatus
from app.schemas.customer import CustomerInfo
from app.schemas.employee import EmployeeInfo
from app.schemas.shipper import ShipperInfo

class OrderDetailCreate(BaseModel):
    product_id: int
    quantity: int = Field(..., gt=0)
    unit_price: Optional[Decimal] = None  # If not provided, use product price
    discount: Decimal = Field(default=Decimal(0), ge=0, le=1)

class OrderCreate(BaseModel):
    customer_id: str
    employee_id: Optional[int] = None
    order_date: Optional[date] = None
    required_date: Optional[date] = None
    ship_via: Optional[int] = None
    freight: Decimal = Field(default=Decimal(0), ge=0)
    ship_name: Optional[str] = None
    ship_address: Optional[str] = None
    ship_city: Optional[str] = None
    ship_region: Optional[str] = None
    ship_postal_code: Optional[str] = None
    ship_country: Optional[str] = None
    order_details: List[OrderDetailCreate] = Field(..., min_length=1)

class OrderUpdate(BaseModel):
    customer_id: Optional[str] = None
    employee_id: Optional[int] = None
    order_date: Optional[date] = None
    required_date: Optional[date] = None
    shipped_date: Optional[date] = None
    ship_via: Optional[int] = None
    freight: Optional[Decimal] = None
    ship_name: Optional[str] = None
    ship_address: Optional[str] = None
    ship_city: Optional[str] = None
    ship_region: Optional[str] = None
    ship_postal_code: Optional[str] = None
    ship_country: Optional[str] = None
    order_details: Optional[List[OrderDetailCreate]] = None

class OrderDetailResponse(BaseModel):
    product_id: int
    product_name: Optional[str] = None # Filled from relationship
    unit_price: Decimal
    quantity: int
    discount: Decimal
    line_total: Decimal
    discount_amount: Decimal
    final_total: Decimal
    model_config = ConfigDict(from_attributes=True)

class OrderResponse(BaseModel):
    order_id: int
    customer: Optional[CustomerInfo] = None
    employee: Optional[EmployeeInfo] = None
    shipper: Optional[ShipperInfo] = None
    order_date: Optional[date]
    required_date: Optional[date]
    shipped_date: Optional[date]
    status: OrderStatus
    freight: Decimal
    subtotal: Decimal
    discount_total: Decimal
    total: Decimal
    order_details: List[OrderDetailResponse]
    
    ship_name: Optional[str] = None
    ship_address: Optional[str] = None
    ship_city: Optional[str] = None
    ship_region: Optional[str] = None
    ship_postal_code: Optional[str] = None
    ship_country: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)

class OrderListResponse(BaseModel):
    order_id: int
    customer: Optional[CustomerInfo] = None
    employee: Optional[EmployeeInfo] = None
    order_date: Optional[date]
    required_date: Optional[date]
    shipped_date: Optional[date]
    status: OrderStatus
    total: Decimal
    model_config = ConfigDict(from_attributes=True)
